backend/test/mission_service_impl_test.cpp

```diff
@@ -1,55 +1,358 @@
+}
+
+std::future<void> MissionServiceImplUploadTest::uploadMissionAndSaveParams(
+    std::shared_ptr<UploadMissionRequest> request,
+    std::shared_ptr<UploadMissionResponse> response)
+{
+    EXPECT_CALL(_mission, upload_mission_async(_, _))
+    .WillOnce(SaveUploadParams(&_uploaded_mission, &_result_callback, &_callback_saved_promise));
+
+    auto upload_handle = std::async(std::launch::async, [this, request, response]() {
+        _mission_service.UploadMission(nullptr, request.get(), response.get());
+    });
+
+    _callback_saved_future.wait();
+    return upload_handle;
+}
+
+std::shared_ptr<UploadMissionRequest>
+MissionServiceImplUploadTest::generateUploadRequest(const
+                                                    std::vector<std::shared_ptr<dc::MissionItem>>
+                                                    &mission_items) const
+{
+    auto request = std::make_shared<UploadMissionRequest>();
+    auto rpc_mission = new dc::rpc::mission::Mission();
+
+    for (const auto &mission_item : mission_items) {
+        auto rpc_mission_item = rpc_mission->add_mission_item();
+        rpc_mission_item->set_latitude_deg(mission_item->get_latitude_deg());
+        rpc_mission_item->set_longitude_deg(mission_item->get_longitude_deg());
+        rpc_mission_item->set_relative_altitude_m(mission_item->get_relative_altitude_m());
+        rpc_mission_item->set_speed_m_s(mission_item->get_speed_m_s());
+        rpc_mission_item->set_is_fly_through(mission_item->get_fly_through());
+        rpc_mission_item->set_gimbal_pitch_deg(mission_item->get_gimbal_pitch_deg());
+        rpc_mission_item->set_gimbal_yaw_deg(mission_item->get_gimbal_yaw_deg());
+        rpc_mission_item->set_camera_action(translateCameraAction(mission_item->get_camera_action()));
+    }
+
+    request->set_allocated_mission(rpc_mission);
+
+    return request;
+}
+
+RPCCameraAction MissionServiceImplUploadTest::translateCameraAction(const CameraAction
+                                                                    camera_action) const
+{
+    return static_cast<RPCCameraAction>(camera_action);
+}
+
+TEST_F(MissionServiceImplUploadTest, uploadsEmptyMissionWhenNullRequest)
+{
+    auto upload_handle = uploadMissionAndSaveParams(nullptr, nullptr);
+
+    _result_callback(dc::Mission::Result::UNKNOWN);
+    upload_handle.wait();
+
+    EXPECT_EQ(0, _uploaded_mission.size());
 }
 
+TEST_F(MissionServiceImplUploadTest, uploadsOneItemMission)
+{
+    auto mission_items = generateListOfOneItem();
+    checkItemsAreUploadedCorrectly(mission_items);
+}
+
+std::vector<std::shared_ptr<dc::MissionItem>> MissionServiceImplUploadTest::generateListOfOneItem()
+{
+    std::vector<std::shared_ptr<dc::MissionItem>> mission_items;
+
+    auto mission_item = std::make_shared<dc::MissionItem>();
+    mission_item->set_position(41.848695, 75.132751);
+    mission_item->set_relative_altitude(50.4);
+    mission_item->set_speed(8.3);
```

Comment:
Some of the fields are actually floats and so it should be `8.3f`, however, I wonder if we shouldn't just make everything `double` in DroneCore, especially in the public interface since DroneCore is not run on embedded platforms generally and everything is becoming 64 bits these days.